presto-common/src/main/java/com/facebook/presto/common/block/MapBlock.java

```diff
@@ -0,0 +1,303 @@
+            @Nullable boolean[] mapIsNull,
+            int[] offsets,
+            Block keyBlock,
+            Block valueBlock,
+            HashTables hashTables)
+    {
+        int[] rawHashTables = hashTables.get();
+        if (rawHashTables != null && rawHashTables.length < keyBlock.getPositionCount() * HASH_MULTIPLIER) {
+            throw new IllegalArgumentException(format("keyBlock/valueBlock size does not match hash table size: %s %s", keyBlock.getPositionCount(), rawHashTables.length));
+        }
+
+        this.startOffset = startOffset;
+        this.positionCount = positionCount;
+        this.mapIsNull = mapIsNull;
+        this.offsets = offsets;
+        this.keyBlock = keyBlock;
+        this.valueBlock = valueBlock;
+        this.hashTables = hashTables;
+        this.sizeInBytes = -1;
+        this.logicalSizeInBytes = -1;
+
+        // We will add the hashtable size to the retained size even if it's not built yet. This could be overestimating
+        // but is necessary to avoid reliability issues. Currently the memory counting framework only pull the retained
+        // size once for each operator so updating in the middle of the processing would not work.
+        this.retainedSizeInBytesExceptHashtable = INSTANCE_SIZE
+                + keyBlock.getRetainedSizeInBytes()
+                + valueBlock.getRetainedSizeInBytes()
+                + sizeOf(offsets)
+                + sizeOf(mapIsNull);
+    }
+
+    @Override
+    protected Block getRawKeyBlock()
+    {
+        return keyBlock;
+    }
+
+    @Override
+    protected Block getRawValueBlock()
+    {
+        return valueBlock;
+    }
+
+    @Override
+    protected HashTables getHashTables()
+    {
+        return hashTables;
+    }
+
+    @Override
+    protected int[] getOffsets()
+    {
+        return offsets;
+    }
+
+    @Override
+    public int getOffsetBase()
+    {
+        return startOffset;
+    }
+
+    @Override
+    @Nullable
+    protected boolean[] getMapIsNull()
+    {
+        return mapIsNull;
+    }
+
+    @Override
+    public int getPositionCount()
+    {
+        return positionCount;
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        if (sizeInBytes < 0) {
+            calculateSize();
+        }
+        return sizeInBytes;
+    }
+
+    private boolean isSinglePositionBlock(int position)
+    {
+        return position == 0 && positionCount == 1 && offsets.length == 2;
+    }
+
+    @Override
+    public Block getSingleValueBlock(int position)
+    {
+        if (isSinglePositionBlock(position)) {
+            return this;
+        }
+
+        return getSingleValueBlockInternal(position);
+    }
+
+    private void calculateSize()
+    {
+        int entriesStart = offsets[startOffset];
+        int entriesEnd = offsets[startOffset + positionCount];
+        int entryCount = entriesEnd - entriesStart;
+        sizeInBytes = keyBlock.getRegionSizeInBytes(entriesStart, entryCount) +
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).